apps/indexer/src/services/siwe-auth-service.ts

```diff
@@ -0,0 +1,713 @@
+      const { chainId, address } = siweMessagePayloadSchema.parse(
+        parseSiweMessage(message),
+      );
+
+      if (!chainId) {
+        throw new SiweAuthService_MissingChainIdError();
+      }
+
+      const client = await this.options.resolveChainClient(chainId);
+
+      if (!client) {
+        throw new SiweAuthService_UnsupportedChainError();
+      }
+
+      const verificationResult = await client.verifySiweMessage({
+        message,
+        signature,
+        nonce,
+      });
+
+      if (!verificationResult) {
+        throw new SiweAuthService_SiweMessageVerificationFailedError();
+      }
+
+      return await this.options.db.transaction(async (tx) => {
+        // is there existing auth method for this ethereum address?
+        let authCredentials = await tx.query.userAuthCredentials.findFirst({
+          where(fields, operators) {
+            return operators.and(
+              operators.eq(fields.method, "siwe"),
+              operators.eq(fields.identifier, address),
+            );
+          },
+        });
+
+        if (!authCredentials) {
+          const [user] = await tx
+            .insert(schema.user)
+            .values({})
+            .returning()
+            .execute();
+
+          if (!user) {
+            throw new SiweAuthService_FailedToCreateUserError();
+          }
+
+          // create credentials and account
+          [authCredentials] = await tx
+            .insert(schema.userAuthCredentials)
+            .values({ identifier: address, method: "siwe", userId: user.id })
+            .returning()
+            .execute();
+        }
+
+        if (!authCredentials) {
+          throw new SiweAuthService_FailedToCreateAuthCredentialsError();
+        }
+
+        // create session
+        const [session] = await tx
+          .insert(schema.userAuthSession)
+          .values({
+            userAuthCredentialsId: authCredentials.id,
+            userId: authCredentials.userId,
+          })
+          .returning()
+          .execute();
+
+        if (!session) {
+          throw new SiweAuthService_FailedToCreateAuthSessionError();
+        }
+
+        const now = Math.floor(Date.now() / 1000);
+        const refreshTokenExpiresAt =
+          now + this.options.jwtRefreshTokenLifetime;
+
+        const [storedRefreshToken] = await tx
+          .insert(schema.userAuthSessionSiweRefreshToken)
+          .values({
+            expiresAt: new Date(refreshTokenExpiresAt * 1000),
+            userAuthSessionId: session.id,
+          })
+          .returning()
+          .execute();
+
+        if (!storedRefreshToken) {
+          throw new SiweAuthService_FailedToCreateRefreshTokenError();
+        }
+
+        return {
+          accessToken: await this.createAccessToken(session.id, now),
+          refreshToken: await this.createRefreshToken(
+            storedRefreshToken.id,
+            now,
+          ),
+        };
+      });
+    } catch (e) {
+      throwAuthServiceError(e);
```

Comment:
wrapping the error will bury the source line of code that breaks, making it difficult to see where line of code actually went wrong in debug tools, unless we want to handle the exact type of error, why not just allow the original error pop?